sa/sa.go

```diff
@@ -1493,3 +1493,131 @@ func (ssa *SQLStorageAuthority) UnpauseAccount(ctx context.Context, req *sapb.Re
 
 	return total, nil
 }
+
+// AddRateLimitOverride adds a rate limit override to the database. Only the
+// comment field is optional. If the override already exists, it will be
+// updated. If the override does not exist, it will be inserted and enabled. If
+// the override exists but has been disabled, it will be updated but not be
+// re-enabled. The status of the override is returned in Enabled field of the
+// response. To re-enable an override, use the EnableRateLimitOverride method.
+func (ssa *SQLStorageAuthority) AddRateLimitOverride(ctx context.Context, req *sapb.AddRateLimitOverrideRequest) (*sapb.AddRateLimitOverrideResponse, error) {
+	if core.IsAnyNilOrZero(req, req.Override, req.Override.LimitEnum, req.Override.BucketKey, req.Override.Count, req.Override.Burst, req.Override.Period) {
+		return nil, errIncompleteRequest
+	}
+
+	var inserted bool
+	var enabled bool
+	now := ssa.clk.Now()
+
+	_, err := db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (any, error) {
+		var existing overrideModel
+		err := tx.SelectOne(ctx, &existing, `
+			SELECT createdAt, enabled
+			  FROM overrides
+			 WHERE limitEnum = ? AND
+			       bucketKey = ?`,
+			req.Override.LimitEnum,
+			req.Override.BucketKey,
+		)
+
+		switch {
+		case err != nil && !db.IsNoRows(err):
+			// Error querying the database.
+			return nil, fmt.Errorf("querying rate limit override: %w", err)
+
+		case db.IsNoRows(err):
+			// Insert a new overrides row.
+			new := newOverrideModelFromPB(req.Override, now, now, true)
+			err = tx.Insert(ctx, &new)
+			if err != nil && !db.IsDuplicate(err) {
```

Comment:
And if the error is a duplicate row error? I believe that's technically possible despite our transaction here, because if the row doesn't exist, the Select won't have locked anything and another thread could come in and create a row with the same primary key. But that row might not have the same data as the one we're trying to insert here! I think we should probably return an error rather than just eliding it.